src/split.cc

```diff
@@ -0,0 +1,155 @@
+	const std::array<size_t, 1> actual_num_chunks = {std::min(num_chunks, full_chunk.range[0] / granularity[0])};
+	const auto [small_chunk_size, large_chunk_size, num_large_chunks] = compute_small_and_large_chunks<1>(full_chunk, granularity, actual_num_chunks);
+
+	std::vector<chunk<3>> result(actual_num_chunks[0], {full_chunk.offset, full_chunk.range, full_chunk.global_size});
+	for(auto i = 0u; i < num_large_chunks[0]; ++i) {
+		result[i].range[0] = large_chunk_size[0];
+		result[i].offset[0] += i * large_chunk_size[0];
+	}
+	for(auto i = num_large_chunks[0]; i < actual_num_chunks[0]; ++i) {
+		result[i].range[0] = small_chunk_size[0];
+		result[i].offset[0] += num_large_chunks[0] * large_chunk_size[0] + (i - num_large_chunks[0]) * small_chunk_size[0];
+	}
+
+#ifndef NDEBUG
+	sanity_check_split(full_chunk, result);
+#endif
+
+	return result;
+}
+
+// TODO: Make the split dimensions configurable for 3D chunks?
+std::vector<chunk<3>> split_2d(const chunk<3>& full_chunk, const range<3>& granularity, const size_t num_chunks) {
+#ifndef NDEBUG
+	assert(num_chunks > 0);
+	for(int d = 0; d < 3; ++d) {
+		assert(granularity[d] > 0);
+		assert(full_chunk.range[d] % granularity[d] == 0);
+	}
+#endif
+
+	const auto assign_factors = [&full_chunk, &granularity, &num_chunks](const size_t factor) {
+		assert(num_chunks % factor == 0);
+		const size_t max_chunks[2] = {full_chunk.range[0] / granularity[0], full_chunk.range[1] / granularity[1]};
+		const size_t f0 = factor;
+		const size_t f1 = num_chunks / factor;
+
+		// Decide in which direction to split by first checking which
+		// factor assignment produces more chunks under the given constraints.
+		const std::array<size_t, 2> split0 = {std::min(f0, max_chunks[0]), std::min(f1, max_chunks[1])};
+		const std::array<size_t, 2> split1 = {std::min(f1, max_chunks[0]), std::min(f0, max_chunks[1])};
+		const auto count0 = split0[0] * split0[1];
+		const auto count1 = split1[0] * split1[1];
+
+		if(count0 == count1) {
+			// If we're tied for the number of chunks we can create, try some heuristics to decide.
+
+			// If domain is square(-ish), prefer splitting along slower dimension.
+			// (These bounds have been chosen arbitrarily!)
+			const double squareishness = std::sqrt(full_chunk.range.size()) / static_cast<double>(full_chunk.range[0]);
+			if(squareishness > 0.95 && squareishness < 1.05) { return (f0 >= f1) ? split0 : split1; }
+
+			// For non-square domains, prefer split that produces shorter edges (compare sum of circumferences)
+			const auto circ0 = full_chunk.range[0] / split0[0] + full_chunk.range[1] / split0[1];
+			const auto circ1 = full_chunk.range[0] / split1[0] + full_chunk.range[1] / split1[1];
+			return circ0 < circ1 ? split0 : split1;
+
+			// TODO: Yet another heuristic we may want to consider is how even chunk sizes are,
+			// i.e., how balanced the workload is.
+		}
+		if(count0 > count1) { return split0; }
+		return split1;
```

Comment:
Consider making these two the first branches in an if-else statement (they're the easiest to decide).